### ISSUE TITLE
Performance issues on larger sites

### ISSUE THREAD

**Issue by andrewmenich** [doublesecretagency/craft-upvote#40]
Howdy! We've recently implemented the Upvote plugin on a fairly large site (still in development) and noticed that the template tag `{% do craft.upvote.sort(entryQuery) %}` was causing some  performance issues. The updated `orderBy` clause contained 320k+ IDs, corresponding to the number of elements in the `elements` table, which seemed to be adding a couple of seconds to the page load time. I expect this number of elements will only grow and thus the performance hit will grow as well. 

I tinkered with the underlying function [`_elementIdsByTally()`](https://github.com/doublesecretagency/craft-upvote/blob/6905a11afe975ac0e2a5abf2866cde7d240abe54/src/services/Query.php#L276) that gets the element IDs to sort by and had some moderate success in reducing load times (~3-4x faster) for our use case. 

The first change was to limit the $elementIds query to only get elements by the query's elementType since those are the only IDs that matter anyway

```diff
- private function _elementIdsByTally(?string $key): array
+ private function _elementIdsByTally(?string $key, EntryQuery $query): array
  {
       …
       $elementIds = (new CraftQuery())
            ->select('[[elements.id]]')
            ->from('{{%elements}} elements')
+           ->where(['[[elements.type]]' => $query->elementType])
            ->leftJoin(['subquery' => $subquery], '[[elements.id]] = [[subquery.elementId]]')
            ->orderBy([new Expression($queryOrder)]);
       …
  }
```

This change greatly reduced the amount of IDs getting returned here and thus improved performance.

Another change I played with was cloning the input query and executing it with `->ids()` and using that array to further reduce the number of IDs returned here.

```diff
- private function _elementIdsByTally(?string $key): array
+ private function _elementIdsByTally(?string $key, EntryQuery $query): array
  {
+     $clonedQuery = clone $query;
+     $ids = $clonedQuery->ids();
       …
       $elementIds = (new CraftQuery())
            ->select('[[elements.id]]')
            ->from('{{%elements}} elements')
+           ->where(['[[elements.type]]' => $query->elementType])
+           ->andWhere(['[[elements.id]]' => $ids])
            ->leftJoin(['subquery' => $subquery], '[[elements.id]] = [[subquery.elementId]]')
            ->orderBy([new Expression($queryOrder)]);
       …
  }
```

I understand this approach could yield performance issues as it means executing the primary query twice, but in my case it seemed to improve the performance, especially when the primary query is using `->limit()`. Perhaps the plugin could have a setting to toggle this approach? 

Happy to submit a PR for these changes if they make sense. Cheers!

**Comment by lindseydiloreto** [comment]
This is brilliant, I would happily take a PR with this fix...


```diff
- private function _elementIdsByTally(?string $key): array
+ private function _elementIdsByTally(?string $key, EntryQuery $query): array
  {
       …
       $elementIds = (new CraftQuery())
            ->select('[[elements.id]]')
            ->from('{{%elements}} elements')
+           ->where(['[[elements.type]]' => $query->elementType])
            ->leftJoin(['subquery' => $subquery], '[[elements.id]] = [[subquery.elementId]]')
            ->orderBy([new Expression($queryOrder)]);
       …
  }
```

The other one (cloning the query) is a hair more debatable, but probably still a solid improvement in 90% of cases.

Let's look at it this way... for the small number of cases where a double-query is detrimental, it probably means they don't have an overwhelming number of votes, and thus are unlikely to be dealing that that (much more significant) problem.

So **yes please, I'll take a PR for both!** Thank you so much @andrewmenich, these are great improvements. 🙏

**Comment by andrewmenich** [comment]
Thanks! PR is open. Happy to kill off the second bit and just go with adding the elementType to the query if that's more agreeable!

**Comment by lindseydiloreto** [comment]
Thanks for this suggestion and PR, in theory it should speed things up for everyone!

This change is now available in [v2.3.3](https://github.com/doublesecretagency/craft-upvote/releases/tag/2.3.3). 👍
